pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -623,338 +671,284 @@ func (r *Reconciler) shouldCreateNewResourceSnapshotNow(ctx context.Context, lat
-		j := i
-		currentSize := 0
-		var snapshotResources []fleetv1beta1.ResourceContent
-		for j < len(selectedResources) {
-			currentSize += len(selectedResources[j].Raw)
-			if currentSize > resourceSnapshotResourceSizeLimit {
-				break
-			}
-			snapshotResources = append(snapshotResources, selectedResources[j])
-			j++
-		}
-		// Any selected resource will always be less than 1.5MB since that's the ETCD limit. In this case an individual
-		// selected resource crosses the 1MB limit.
-		if len(snapshotResources) == 0 && len(selectedResources[j].Raw) > resourceSnapshotResourceSizeLimit {
-			snapshotResources = append(snapshotResources, selectedResources[j])
-			j++
-		}
-		selectedResourcesList = append(selectedResourcesList, snapshotResources)
-		i = j
-	}
-	return selectedResourcesList
-}
-
-// ensureLatestPolicySnapshot ensures the latest policySnapshot has the isLatest label and the numberOfClusters are updated.
-func (r *Reconciler) ensureLatestPolicySnapshot(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement, latest *fleetv1beta1.ClusterSchedulingPolicySnapshot) error {
+// ensureLatestPolicySnapshot ensures the latest policySnapshot has the isLatest label and the numberOfClusters are updated for interface types.
+func (r *Reconciler) ensureLatestPolicySnapshot(ctx context.Context, placementObj fleetv1beta1.PlacementObj, latest fleetv1beta1.PolicySnapshotObj) error {
 	needUpdate := false
 	latestKObj := klog.KObj(latest)
-	if latest.Labels[fleetv1beta1.IsLatestSnapshotLabel] != strconv.FormatBool(true) {
+	labels := latest.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	if labels[fleetv1beta1.IsLatestSnapshotLabel] != strconv.FormatBool(true) {
 		// When latestPolicySnapshot.Spec.PolicyHash == policyHash,
 		// It could happen when the controller just sets the latest label to false for the old snapshot, and fails to
 		// create a new policy snapshot.
 		// And then the customers revert back their policy to the old one again.
 		// In this case, the "latest" snapshot without isLatest label has the same policy hash as the current policy.
 
-		latest.Labels[fleetv1beta1.IsLatestSnapshotLabel] = strconv.FormatBool(true)
+		labels[fleetv1beta1.IsLatestSnapshotLabel] = strconv.FormatBool(true)
+		latest.SetLabels(labels)
 		needUpdate = true
 	}
 	crpGeneration, err := annotations.ExtractObservedCRPGenerationFromPolicySnapshot(latest)
 	if err != nil {
-		klog.ErrorS(err, "Failed to parse the CRPGeneration from the annotations", "clusterSchedulingPolicySnapshot", latestKObj)
+		klog.ErrorS(err, "Failed to parse the CRPGeneration from the annotations", "policySnapshot", latestKObj)
 		return controller.NewUnexpectedBehaviorError(err)
 	}
-	if crpGeneration != crp.Generation {
-		latest.Annotations[fleetv1beta1.CRPGenerationAnnotation] = strconv.FormatInt(crp.Generation, 10)
+	if crpGeneration != placementObj.GetGeneration() {
+		annotations := latest.GetAnnotations()
+		if annotations == nil {
```

Comment:
ditto, fleetv1beta1.CRPGenerationAnnotation: strconv.FormatInt(crp.Generation, 10) annoation is required for each snapshot. so nil is impossible